flutter/flutter
# Update ink feature splashFactory test coverage to not rely on the Theme's default value

## Description

This is the first in a series of PRs planned for several small improvements to the `InkWell` and its ripple animations. See https://github.com/flutter/flutter/issues/73163 for additional context:

> SplashFactory Test Coverage: Currently, many tests fail after changing the default splash factory on the material Theme. To make the PR changing the default splash factory easier, test coverage should be improved for widgets using ink features, to be able to pass regardless of what the theme’s default splash factory is. We can take the opportunity here to consolidate some duplicated code for utils related to testing how ripples are drawn.

This PR is just updating tests for widgets using `InkWell`s, to include test coverage for when the `InkWell`s in question will use both the `InkRipple.splashFactory` and the `InkSplash.splashFactory`. Specifically, this PR is meant to improve test coverage, so that when we update the default `splashFactory` on the material `Theme` from the `InkSplash.splashFactory` to the `InkRipple.splashFactory`, it isn't necessary to update a bunch of failing tests, but just add a small test for the default value on the `Theme`.

Also, you'll find that I added some small utils in `ink_paint_test_utils.dart`, to help consolidate a bunch of boilerplate code. This contributed to some extra diffs. Similarly, in places where it made sense, I organized tests into groups, which similarly led to a bunch of extra diffs.

## Related Issues

- [#73163: [proposal] InkRipple and InkSplash implementations warrant improvements](https://github.com/flutter/flutter/issues/73163)

## Tests

I added/modified the following tests:
`bottom_navigation_bar_test`:
- **BottomNavigationBar shifting backgroundColor with transition**

`chip_test`
- _Ink features
  - with InkSplashFactory_
    - **Chip creates centered, unique splash when label is tapped**
    - **Delete button creates non-centered, unique splash when tapped**
    - **Chip without delete button creates correct ripple**
    - **Selection with avatar works as expected on RawChip**
    - **Selection without avatar works as expected on RawChip**
    - **Activation works as expected on RawChip**
  - _with InkRippleFactory_
    - **Chip creates centered, unique splash when label is tapped**
    - **Delete button creates non-centered, unique splash when tapped**
    - **Chip without delete button creates correct ripple**
    - **Selection with avatar works as expected on RawChip**
    - **Selection without avatar works as expected on RawChip**
    - **Activation works as expected on RawChip**

`data_table_test`
- _DataRow custom colors_
  - **DataRow renders custom colors when selected**
  - **DataRow renders custom colors when disabled**
  - **DataRow with InkSplash splash factory renders custom colors when pressed**
  - **DataRow with InkRipple splash factory renders custom colors when pressed**

`icon_button_test`
- _Ink features_
  - **IconButton with explicit splashColor and highlightColor**
  - _with InkSplash splash factory_
    - **IconButton draws splash with explicit splashRadius**
  - _with InkRipple splash factory_
    - **IconButton ripple radius is 5 greater than splashRadius**

`ink_paint_test`
- **The InkWell widget renders an ink splash**
- **Does the Ink widget render anything**
- **The InkWell widget renders an SelectAction or ActivateAction-induced ink ripple**
- **Canceling InkRipple that was disposed when its animation ended draws one transparent circle**

`ink_well_test` (a bunch of organizing)
- _Ink feature color states_
  - _on hover_
    - **InkWell uses hoverColor**
    - **InkWell uses resolved overlayColor MaterialState.hovered**
  - _on focus_
    - **ink response changes color on focus**
    - **ink response changes color on focus with overlayColor**
    - **ink response uses radius for focus highlight**
    - **ink response doesn't change color on focus when on touch device**
  - _on pressed_
    - _with InkSplash splash factory_
      - **splash color matches splashColor**
      - **splash color matches resolved overlayColor for MaterialState.pressed**
    - _with InkSplash splash factory_
      - **ripple color matches splashColor**
      - **ripple color matches resolved overlayColor for MaterialState.pressed**

`raw_material_button_test`
- _Ink Features_
  - _with InkSplash.splashFactory_
    - **RawMaterialButton responds with splash when tapped**
    - **RawMaterialButton responds to shortcut with splash when activated**
    - **Ink splash from center tap originates in correct location**
    - **Ink splash from tap above material originates in correct location**
  - _with InkRipple.splashFactory_
    - **RawMaterialButton responds with ripple when tapped**
    - **RawMaterialButton responds to shortcut with ripple when activated**
    - **Ink ripple from center tap originates in correct location**
    - **Ink ripple from tap above material originates in correct location**

`tabs_test`
- **Tab's InkSplash color matches resolved Tab overlayColor for MaterialState.pressed**
- **Tab's InkRipple color matches resolved Tab overlayColor for MaterialState.pressed**

*Replace this with a list of the tests that you added as part of this PR. A change in behavior with no test covering it
will likely get reverted accidentally sooner or later. PRs must include tests for all changed/updated/fixed behaviors. See [Test Coverage].*

## Checklist

Before you create this PR, confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.


## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].



[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes
